app_dart/test/src/bigquery/fake_tabledata_resource.dart

```diff
@@ -0,0 +1,37 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:googleapis/bigquery/v2.dart';
+
+/// A fake bigquery tabledataresourceApi implementation.
+///
+/// This tabledataResourceApi considers only a simple case
+/// where we focus on the number of rows inserted. This can be
+/// easily extended for other test cases.
+class FakeTabledataResourceApi implements TabledataResourceApi {
+  int rows = 0;
+  
+  @override
+  Future<TableDataInsertAllResponse> insertAll(
+      TableDataInsertAllRequest request,
+      String projectId,
+      String datasetId,
+      String tableId,
+      {String $fields}) async {
+    rows += request.rows.length;
+    return TableDataInsertAllResponse.fromJson(<String, String>{});
+  }
+
+  @override
+  Future<TableDataList> list(String projectId, String datasetId, String tableId,
+      {int maxResults,
+      String selectedFields,
+      String startIndex,
+      String pageToken,
+      String $fields}) async {
+    return TableDataList.fromJson(
+        <String, String>{'totalRows': rows.toString()});
+  }
+}
+
```

Comment:
Line 22-23 and line 34 are the implementations, though they are simple. You can refer to the case of datastore. They are only used to mimic real API to let test continue. As for how complex our implement should be, then it depends on our real purpose of targeted features.

Reply:
Then `list` can be applied to avoid exposing the implementation details.

Reply:
Not sure what you mean. Can you elaborate?

Reply:
As `insertAll` writes data and `list` reads data, a test could utilize these original API surfaces without interacting with an implementation detail (`row`).

Reply:
What you are suggesting is good if we can directly use the original APIs. However, they are not applicable in the test case since they communicate with real BigQuery APIs. And that's why we have to implement the fake, which mimics functionalities of real APIs. And here comes the 'row'..

Reply:
What you are suggesting is good if we can directly use the original APIs. However, they are not applicable in the test case since they communicate with real BigQuery APIs. And that's why we have to implement the fake, which mimics functionalities of real APIs. And here comes the 'row'..